# 01. Importing libraries

In [30]:
import pandas as pd
import numpy as np
import os 

# 02. Importing data

In [31]:
# create path
path = r'/Users/tuyetnguyen/Desktop/Career Foundry/Achievement 4/04-2023 Instacart Basket Analysis'

In [32]:
# import products.csv 
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [33]:
# import orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 03. Mixed-Type Data


In [34]:
# create a dataframe
df_test = pd.DataFrame()

In [35]:
# create a mixed-type column
df_test['mix'] = ['a', 'b', 1, True]

In [36]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [37]:
df_test['mix'].dtype

dtype('O')

In [38]:
# check for mixed type
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [39]:
# change data type to str
df_test['mix'] = df_test['mix'].astype('str')

In [40]:
df_test['mix'].dtypes

dtype('O')

# Missing Values


In [41]:
# find missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [42]:
# create subset of the dataframe containing only the values in question
df_nan = df_prods[df_prods['product_name'].isnull()== True]

In [43]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

###### A few ways to deal with missing data:
        1. Create a new variable that acts like a flag based on the missing value.
        2. Impute the value with the mean or median of the column (if the variable is numeric).
        - Mean: df['column with missings'].fillna(mean value, inplace=True)
        - Median: df['column with missings'].fillna(median value, inplace=True)

        3. Remove or filter out the missing data.
        - All missing values: df_prods.dropna(inplace = True)
        - Missing values in a particular col: 
            df_prods.dropna(subset = [‘product_name’], inplace = True)
        


In [44]:
df_prods.shape

(49693, 5)

In [45]:
# create a new df without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [46]:
df_prods_clean.shape

(49677, 5)

# 05. Duplicates

In [47]:
# find duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [48]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [49]:
df_prods_clean.shape

(49677, 5)

In [50]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()


In [51]:
df_prods_clean_no_dups.shape

(49672, 5)

# 06. Tidying Up and Exporting Changes

In [52]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task

###### 1. If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now.

###### 2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
Tip: Keep an eye on min and max values!

In [53]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

###### Nothing looks off

##### 3. Check for mixed-type data in your df_ords dataframe.


In [54]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

No mixed-type data found

##### 4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.
No mixed-type data found

##### 5. Run a check for missing values in your df_ords dataframe.
In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [55]:
# find missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206209 missing values in column days_since_prior_order

In [56]:
# create a df for missing values for further investigation
df_ords_nan = df_ords [df_ords ['days_since_prior_order'].isnull () == True]

In [57]:
df_ords_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 8 columns]

The number of missing values matches the user_id count as well as the latest the user_id. The missing values in the column "days_since_prior_order" are likely due to the fact that these customers are new to the store and have not yet made a prior purchase.

In [59]:
crosstab = pd.crosstab(df_ords['days_since_prior_order'], df_ords['order_number'], dropna = False)

In [60]:
crosstab.to_clipboard()

In [62]:
crosstab

order_number            1      2      3      4      5      6      7      8    \
days_since_prior_order                                                         
0.0                       0   2899   2915   2964   2634   2443   2236   2101   
1.0                       0   4822   4613   4690   4423   4047   4000   3744   
2.0                       0   6203   6209   6231   5789   5679   5274   5090   
3.0                       0   7411   7577   7414   7097   6770   6424   6078   
4.0                       0   8415   8476   8469   8029   7608   7092   6735   
5.0                       0   9216   9329   9220   8600   8228   7703   7270   
6.0                       0  11419  11701  11601  10771  10054   9521   8842   
7.0                       0  16788  16609  16341  15135  13834  13268  12452   
8.0                       0  10422  10304  10148   9378   8738   8068   7594   
9.0                       0   7053   7160   7013   6610   5993   5497   5082   
10.0                      0   5969   6037   5938   5376   5018   4479   4245   
11.0                      0   5352   5388   5325   4718   4363   4087   3767   
12.0                      0   5206   5357   5348   4657   4333   3954   3568   
13.0                      0   6098   6265   6003   5520   4855   4480   4083   
14.0                      0   7680   7529   7577   6951   6004   5549   4976   
15.0                      0   5219   5121   5225   4588   4068   3670   3384   
16.0                      0   3791   3794   3690   3316   2994   2665   2370   
17.0                      0   3131   3241   3226   2823   2546   2152   2039   
18.0                      0   3041   2971   2904   2660   2337   2078   1850   
19.0                      0   2861   2999   2877   2579   2359   2081   1872   
20.0                      0   3411   3380   3427   2953   2593   2355   2113   
21.0                      0   4258   4205   4067   3459   3263   2801   2429   
22.0                      0   2956   2866   2887   2504   2168   2015   1756   
23.0                      0   2268   2296   2204   1889   1666   1516   1282   
24.0                      0   1875   1960   1900   1754   1496   1334   1176   
25.0                      0   1843   1879   1847   1619   1400   1198   1016   
26.0                      0   1872   1890   1760   1619   1442   1238   1084   
27.0                      0   2177   2236   2258   1994   1598   1439   1340   
28.0                      0   2728   2760   2797   2287   2035   1745   1592   
29.0                      0   1986   2012   1971   1711   1489   1259   1113   
30.0                      0  47839  47130  48887  38780  31212  25290  20575   

order_number              9      10   ...  91   92   93   94   95   96   97   \
days_since_prior_order                ...                                      
0.0                      1972   1782  ...  156  154  125  151  151  153  120   
1.0                      3503   3358  ...  397  396  373  359  366  330  356   
2.0                      4914   4705  ...  416  405  407  389  359  351  359   
3.0                      5750   5663  ...  345  287  330  294  291  292  254   
4.0                      6534   6395  ...  221  238  225  191  193  185  183   
5.0                      6828   6563  ...  142  128  105  119  119  120  100   
6.0                      8527   7727  ...   77   91   86   82   67   59   58   
7.0                     11612  10980  ...   60   56   48   56   36   37   37   
8.0                      6894   6511  ...   29   22   24   26   23   30   18   
9.0                      4760   4404  ...   12    8   14   12    8    6    7   
10.0                     3844   3531  ...    9   20   10    6   11    1    7   
11.0                     3429   3180  ...   13   13    5    7    8    6    6   
12.0                     3404   2941  ...    4    5    4    5    6    5    5   
13.0                     3708   3408  ...    8    9    5    6    5    2    3   
14.0                     4563   4187  ...    1    3   12    3    3    3    4   

Here, if you look at where your “days_since_prior_order” column crosses with order numbers of 1, you’ll see that the entire column is populated with 0s. This supports your initial assumption about missing values and means you can safely disregard them.
We found missing values in this column but conducted an analysis that revealed them to be fine in context. Please confirm whether our assumption is correct.”

##### 6. Address the missing values using an appropriate method.
In a markdown cell, explain why you used your method of choice.

The analyst decided to keep the missing values in the data as they may contain meaningful information.

##### 7. Run a check for duplicate values in your df_ords data.
In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [ ]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [ ]:
df_ords_dups

##### 8. Address the duplicates using an appropriate method.
In a markdown cell, explain why you used your method of choice.

There're no duplicates found. No action needed.

##### 9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [ ]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))